File: sdk/core/android/src/main/kotlin/mediation/PriorityWeightedMediation.kt
package com.rivalapex.mediation

import kotlinx.coroutines.*
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.atomic.AtomicInteger
import java.util.concurrent.atomic.AtomicLong
import kotlin.math.max
import kotlin.math.min
import kotlin.random.Random

/**
 * Ad source with priority and weight configuration
 */
data class WeightedAdSource(
    val id: String,
    val priority: Int,
    val weight: Double = 1.0,
    val timeout: Long = 5000L,
    val enabled: Boolean = true,
    val minBid: Double = 0.0,
    val metadata: Map<String, String> = emptyMap()
) {
    init {
        require(priority >= 0) { "Priority must be non-negative" }
        require(weight > 0) { "Weight must be positive" }
        require(timeout > 0) { "Timeout must be positive" }
    }
}

/**
 * Result of loading an ad from a source
 */
sealed class AdLoadResult {
    data class Success(val ad: Any, val sourceId: String, val latencyMs: Long, val bid: Double = 0.0) : AdLoadResult()
    data class NoFill(val sourceId: String, val reason: String) : AdLoadResult()
    data class Error(val sourceId: String, val error: Throwable) : AdLoadResult()
    data class Timeout(val sourceId: String) : AdLoadResult()
}

/**
 * Statistics for a source
 */
data class SourcePerformance(
    val sourceId: String,
    val totalAttempts: Int,
    val successCount: Int,
    val noFillCount: Int,
    val errorCount: Int,
    val timeoutCount: Int,
    val averageLatencyMs: Double,
    val fillRate: Double,
    val averageBid: Double,
    val effectiveWeight: Double
)

/**
 * Configuration for priority-weighted mediation
 */
data class PriorityWeightedConfig(
    val usePerformanceWeighting: Boolean = true,
    val performanceWindowMs: Long = 3_600_000, // 1 hour
    val minSampleSize: Int = 10,
    val weightDecayFactor: Double = 0.1,
    val maxConcurrentRequests: Int = 3,
    val bidFloorEnabled: Boolean = true,
    val adaptiveTimeoutsEnabled: Boolean = true
)

/**
 * Ad loader function type
 */
typealias AdLoader = suspend (source: WeightedAdSource) -> AdLoadResult

/**
 * Priority-weighted mediation manager
 *
 * Implements intelligent ad source selection based on:
 * - Configured priority (lower = higher priority)
 * - Weighted random selection within same priority tier
 * - Performance-based weight adjustments
 * - Fill rate optimization
 */
class PriorityWeightedMediation(
    private val config: PriorityWeightedConfig = PriorityWeightedConfig()
) {
    private val sourceStats = ConcurrentHashMap<String, SourceStats>()
    private val scope = CoroutineScope(Dispatchers.IO + SupervisorJob())
    
    private class SourceStats {
        val attempts = AtomicInteger(0)
        val successes = AtomicInteger(0)
        val noFills = AtomicInteger(0)
        val errors = AtomicInteger(0)
        val timeouts = AtomicInteger(0)
        val totalLatencyMs = AtomicLong(0)
        val totalBid = AtomicLong(0) // Stored as micros
        var lastSuccessTime = 0L
    }
    
    /**
     * Execute mediation with priority-weighted selection
     */
    suspend fun execute(
        sources: List<WeightedAdSource>,
        loader: AdLoader
    ): AdLoadResult {
        if (sources.isEmpty()) {
            return AdLoadResult.NoFill("none", "No sources configured")
        }
        
        val enabledSources = sources.filter { it.enabled }
        if (enabledSources.isEmpty()) {
            return AdLoadResult.NoFill("none", "All sources disabled")
        }
        
        // Group by priority
        val priorityGroups = enabledSources
            .groupBy { it.priority }
            .toSortedMap()
        
        // Try each priority group in order
        for ((_, group) in priorityGroups) {
            val result = executePriorityGroup(group, loader)
            if (result is AdLoadResult.Success) {
                return result
            }
        }
        
        return AdLoadResult.NoFill("all", "All sources exhausted")
    }
    
    /**
     * Execute mediation within a single priority group using weighted selection
     */
    private suspend fun executePriorityGroup(
        sources: List<WeightedAdSource>,
        loader: AdLoader
    ): AdLoadResult {
        val remainingSources = sources.toMutableList()
        
        while (remainingSources.isNotEmpty()) {
            val selected = selectByWeight(remainingSources)
            remainingSources.remove(selected)
            
            val result = executeWithTimeout(selected, loader)
            recordResult(selected.id, result)
            
            if (result is AdLoadResult.Success) {
                return result
            }
        }
        
        return AdLoadResult.NoFill("priority_group", "No fill from priority group")
    }
    
    /**
     * Select a source using weighted random selection
     */
    private fun selectByWeight(sources: List<WeightedAdSource>): WeightedAdSource {
        if (sources.size == 1) {
            return sources[0]
        }
        
        val effectiveWeights = sources.map { source ->
            source to calculateEffectiveWeight(source)
        }
        
        val totalWeight = effectiveWeights.sumOf { it.second }
        if (totalWeight <= 0) {
            return sources.random()
        }
        
        var random = Random.nextDouble() * totalWeight
        for ((source, weight) in effectiveWeights) {
            random -= weight
            if (random <= 0) {
                return source
            }
        }
        
        return sources.last()
    }
    
    /**
     * Calculate effective weight based on configuration and performance
     */
    private fun calculateEffectiveWeight(source: WeightedAdSource): Double {
        val baseWeight = source.weight
        
        if (!config.usePerformanceWeighting) {
            return baseWeight
        }
        
        val stats = sourceStats[source.id] ?: return baseWeight
        
        // Need minimum samples
        if (stats.attempts.get() < config.minSampleSize) {
            return baseWeight
        }
        
        // Calculate fill rate
        val attempts = stats.attempts.get().toDouble()
        val successes = stats.successes.get().toDouble()
        val fillRate = successes / max(1.0, attempts)
        
        // Weight adjustment: boost high performers, penalize low performers
        // Using a sigmoid-like function centered around 50% fill rate
        val performanceMultiplier = 0.5 + fillRate
        
        return max(0.1, baseWeight * performanceMultiplier)
    }
    
    /**
     * Execute ad loading with timeout
     */
    private suspend fun executeWithTimeout(
        source: WeightedAdSource,
        loader: AdLoader
    ): AdLoadResult {
        val timeout = if (config.adaptiveTimeoutsEnabled) {
            calculateAdaptiveTimeout(source)
        } else {
            source.timeout
        }
        
        val startTime = System.currentTimeMillis()
        
        return try {
            withTimeout(timeout) {
                loader(source)
            }
        } catch (e: TimeoutCancellationException) {
            AdLoadResult.Timeout(source.id)
        } catch (e: Exception) {
            AdLoadResult.Error(source.id, e)
        }
    }
    
    /**
     * Calculate adaptive timeout based on historical latency
     */
    private fun calculateAdaptiveTimeout(source: WeightedAdSource): Long {
        val stats = sourceStats[source.id] ?: return source.timeout
        
        val attempts = stats.attempts.get()
        if (attempts < config.minSampleSize) {
            return source.timeout
        }
        
        val avgLatency = stats.totalLatencyMs.get().toDouble() / attempts
        
        // Timeout = 2x average latency, bounded by original timeout
        return min(source.timeout, (avgLatency * 2).toLong().coerceAtLeast(1000L))
    }
    
    /**
     * Record result for statistics
     */
    private fun recordResult(sourceId: String, result: AdLoadResult) {
        val stats = sourceStats.getOrPut(sourceId) { SourceStats() }
        stats.attempts.incrementAndGet()
        
        when (result) {
            is AdLoadResult.Success -> {
                stats.successes.incrementAndGet()
                stats.totalLatencyMs.addAndGet(result.latencyMs)
                stats.totalBid.addAndGet((result.bid * 1_000_000).toLong())
                stats.lastSuccessTime = System.currentTimeMillis()
            }
            is AdLoadResult.NoFill -> {
                stats.noFills.incrementAndGet()
            }
            is AdLoadResult.Error -> {
                stats.errors.incrementAndGet()
            }
            is AdLoadResult.Timeout -> {
                stats.timeouts.incrementAndGet()
            }
        }
    }
    
    /**
     * Get performance statistics for all sources
     */
    fun getPerformanceStats(): List<SourcePerformance> {
        return sourceStats.map { (sourceId, stats) ->
            val attempts = stats.attempts.get()
            val successes = stats.successes.get()
            val noFills = stats.noFills.get()
            val errors = stats.errors.get()
            val timeouts = stats.timeouts.get()
            
            val avgLatency = if (successes > 0) {
                stats.totalLatencyMs.get().toDouble() / successes
            } else 0.0
            
            val fillRate = if (attempts > 0) {
                successes.toDouble() / attempts
            } else 0.0
            
            val avgBid = if (successes > 0) {
                stats.totalBid.get().toDouble() / (successes * 1_000_000)
            } else 0.0
            
            SourcePerformance(
                sourceId = sourceId,
                totalAttempts = attempts,
                successCount = successes,
                noFillCount = noFills,
                errorCount = errors,
                timeoutCount = timeouts,
                averageLatencyMs = avgLatency,
                fillRate = fillRate,
                averageBid = avgBid,
                effectiveWeight = 1.0 // Would need source reference
            )
        }
    }
    
    /**
     * Get performance for a specific source
     */
    fun getSourcePerformance(sourceId: String): SourcePerformance? {
        val stats = sourceStats[sourceId] ?: return null
        val attempts = stats.attempts.get()
        val successes = stats.successes.get()
        
        return SourcePerformance(
            sourceId = sourceId,
            totalAttempts = attempts,
            successCount = successes,
            noFillCount = stats.noFills.get(),
            errorCount = stats.errors.get(),
            timeoutCount = stats.timeouts.get(),
            averageLatencyMs = if (successes > 0) stats.totalLatencyMs.get().toDouble() / successes else 0.0,
            fillRate = if (attempts > 0) successes.toDouble() / attempts else 0.0,
            averageBid = if (successes > 0) stats.totalBid.get().toDouble() / (successes * 1_000_000) else 0.0,
            effectiveWeight = 1.0
        )
    }
    
    /**
     * Reset statistics for a source
     */
    fun resetStats(sourceId: String) {
        sourceStats.remove(sourceId)
    }
    
    /**
     * Reset all statistics
     */
    fun resetAllStats() {
        sourceStats.clear()
    }
    
    /**
     * Cleanup resources
     */
    fun shutdown() {
        scope.cancel()
    }
}

/**
 * Builder for creating weighted ad source configurations
 */
class WeightedAdSourceBuilder {
    private var id: String = ""
    private var priority: Int = 0
    private var weight: Double = 1.0
    private var timeout: Long = 5000L
    private var enabled: Boolean = true
    private var minBid: Double = 0.0
    private var metadata: MutableMap<String, String> = mutableMapOf()
    
    fun id(id: String) = apply { this.id = id }
    fun priority(priority: Int) = apply { this.priority = priority }
    fun weight(weight: Double) = apply { this.weight = weight }
    fun timeout(timeout: Long) = apply { this.timeout = timeout }
    fun enabled(enabled: Boolean) = apply { this.enabled = enabled }
    fun minBid(minBid: Double) = apply { this.minBid = minBid }
    fun metadata(key: String, value: String) = apply { this.metadata[key] = value }
    
    fun build(): WeightedAdSource {
        require(id.isNotEmpty()) { "Source ID is required" }
        return WeightedAdSource(
            id = id,
            priority = priority,
            weight = weight,
            timeout = timeout,
            enabled = enabled,
            minBid = minBid,
            metadata = metadata.toMap()
        )
    }
}

/**
 * Extension function for fluent source creation
 */
fun weightedAdSource(block: WeightedAdSourceBuilder.() -> Unit): WeightedAdSource {
    return WeightedAdSourceBuilder().apply(block).build()
}

File: sdk/core/ios/Sources/Mediation/PriorityWeightedMediation.swift
import Foundation

// MARK: - WeightedAdSource

/// Ad source with priority and weight configuration
public struct WeightedAdSource {
    public let id: String
    public let priority: Int
    public let weight: Double
    public let timeout: TimeInterval
    public let enabled: Bool
    public let minBid: Double
    public let metadata: [String: String]
    
    public init(
        id: String,
        priority: Int,
        weight: Double = 1.0,
        timeout: TimeInterval = 5.0,
        enabled: Bool = true,
        minBid: Double = 0.0,
        metadata: [String: String] = [:]
    ) {
        precondition(priority >= 0, "Priority must be non-negative")
        precondition(weight > 0, "Weight must be positive")
        precondition(timeout > 0, "Timeout must be positive")
        
        self.id = id
        self.priority = priority
        self.weight = weight
        self.timeout = timeout
        self.enabled = enabled
        self.minBid = minBid
        self.metadata = metadata
    }
}

// MARK: - AdLoadResult

/// Result of loading an ad from a source
public enum AdLoadResult {
    case success(ad: Any, sourceId: String, latencyMs: Int64, bid: Double)
    case noFill(sourceId: String, reason: String)
    case error(sourceId: String, error: Error)
    case timeout(sourceId: String)
    
    public var isSuccess: Bool {
        if case .success = self { return true }
        return false
    }
    
    public var sourceId: String {
        switch self {
        case .success(_, let sourceId, _, _): return sourceId
        case .noFill(let sourceId, _): return sourceId
        case .error(let sourceId, _): return sourceId
        case .timeout(let sourceId): return sourceId
        }
    }
}

// MARK: - SourcePerformance

/// Statistics for a source
public struct SourcePerformance {
    public let sourceId: String
    public let totalAttempts: Int
    public let successCount: Int
    public let noFillCount: Int
    public let errorCount: Int
    public let timeoutCount: Int
    public let averageLatencyMs: Double
    public let fillRate: Double
    public let averageBid: Double
    public let effectiveWeight: Double
}

// MARK: - PriorityWeightedConfig

/// Configuration for priority-weighted mediation
public struct PriorityWeightedConfig {
    public var usePerformanceWeighting: Bool
    public var performanceWindowSeconds: TimeInterval
    public var minSampleSize: Int
    public var weightDecayFactor: Double
    public var maxConcurrentRequests: Int
    public var bidFloorEnabled: Bool
    public var adaptiveTimeoutsEnabled: Bool
    
    public init(
        usePerformanceWeighting: Bool = true,
        performanceWindowSeconds: TimeInterval = 3600,
        minSampleSize: Int = 10,
        weightDecayFactor: Double = 0.1,
        maxConcurrentRequests: Int = 3,
        bidFloorEnabled: Bool = true,
        adaptiveTimeoutsEnabled: Bool = true
    ) {
        self.usePerformanceWeighting = usePerformanceWeighting
        self.performanceWindowSeconds = performanceWindowSeconds
        self.minSampleSize = minSampleSize
        self.weightDecayFactor = weightDecayFactor
        self.maxConcurrentRequests = maxConcurrentRequests
        self.bidFloorEnabled = bidFloorEnabled
        self.adaptiveTimeoutsEnabled = adaptiveTimeoutsEnabled
    }
    
    public static let `default` = PriorityWeightedConfig()
}

// MARK: - AdLoader

/// Ad loader function type
public typealias AdLoader = (WeightedAdSource) async throws -> AdLoadResult

// MARK: - PriorityWeightedMediation

/// Priority-weighted mediation manager
public final class PriorityWeightedMediation {
    
    // MARK: - SourceStats
    
    private class SourceStats {
        var attempts: Int = 0
        var successes: Int = 0
        var noFills: Int = 0
        var errors: Int = 0
        var timeouts: Int = 0
        var totalLatencyMs: Int64 = 0
        var totalBidMicros: Int64 = 0
        var lastSuccessTime: Date?
    }
    
    // MARK: - Properties
    
    private let config: PriorityWeightedConfig
    private var sourceStats: [String: SourceStats] = [:]
    private let lock = NSLock()
    
    // MARK: - Initialization
    
    public init(config: PriorityWeightedConfig = .default) {
        self.config = config
    }
    
    // MARK: - Public API
    
    /// Execute mediation with priority-weighted selection
    public func execute(
        sources: [WeightedAdSource],
        loader: @escaping AdLoader
    ) async -> AdLoadResult {
        guard !sources.isEmpty else {
            return .noFill(sourceId: "none", reason: "No sources configured")
        }
        
        let enabledSources = sources.filter { $0.enabled }
        guard !enabledSources.isEmpty else {
            return .noFill(sourceId: "none", reason: "All sources disabled")
        }
        
        // Group by priority
        let priorityGroups = Dictionary(grouping: enabledSources) { $0.priority }
        let sortedPriorities = priorityGroups.keys.sorted()
        
        // Try each priority group in order
        for priority in sortedPriorities {
            guard let group = priorityGroups[priority] else { continue }
            
            let result = await executePriorityGroup(group, loader: loader)
            if result.isSuccess {
                return result
            }
        }
        
        return .noFill(sourceId: "all", reason: "All sources exhausted")
    }
    
    /// Get performance statistics for all sources
    public func getPerformanceStats() -> [SourcePerformance] {
        lock.lock()
        defer { lock.unlock() }
        
        return sourceStats.map { (sourceId, stats) in
            let avgLatency = stats.successes > 0 
                ? Double(stats.totalLatencyMs) / Double(stats.successes) 
                : 0.0
            let fillRate = stats.attempts > 0 
                ? Double(stats.successes) / Double(stats.attempts) 
                : 0.0
            let avgBid = stats.successes > 0 
                ? Double(stats.totalBidMicros) / Double(stats.successes * 1_000_000) 
                : 0.0
            
            return SourcePerformance(
                sourceId: sourceId,
                totalAttempts: stats.attempts,
                successCount: stats.successes,
                noFillCount: stats.noFills,
                errorCount: stats.errors,
                timeoutCount: stats.timeouts,
                averageLatencyMs: avgLatency,
                fillRate: fillRate,
                averageBid: avgBid,
                effectiveWeight: 1.0
            )
        }
    }
    
    /// Get performance for a specific source
    public func getSourcePerformance(sourceId: String) -> SourcePerformance? {
        lock.lock()
        defer { lock.unlock() }
        
        guard let stats = sourceStats[sourceId] else { return nil }
        
        let avgLatency = stats.successes > 0 
            ? Double(stats.totalLatencyMs) / Double(stats.successes) 
            : 0.0
        let fillRate = stats.attempts > 0 
            ? Double(stats.successes) / Double(stats.attempts) 
            : 0.0
        let avgBid = stats.successes > 0 
            ? Double(stats.totalBidMicros) / Double(stats.successes * 1_000_000) 
            : 0.0
        
        return SourcePerformance(
            sourceId: sourceId,
            totalAttempts: stats.attempts,
            successCount: stats.successes,
            noFillCount: stats.noFills,
            errorCount: stats.errors,
            timeoutCount: stats.timeouts,
            averageLatencyMs: avgLatency,
            fillRate: fillRate,
            averageBid: avgBid,
            effectiveWeight: 1.0
        )
    }
    
    /// Reset statistics for a source
    public func resetStats(sourceId: String) {
        lock.lock()
        defer { lock.unlock() }
        sourceStats.removeValue(forKey: sourceId)
    }
    
    /// Reset all statistics
    public func resetAllStats() {
        lock.lock()
        defer { lock.unlock() }
        sourceStats.removeAll()
    }
    
    // MARK: - Private Methods
    
    private func executePriorityGroup(
        _ sources: [WeightedAdSource],
        loader: @escaping AdLoader
    ) async -> AdLoadResult {
        var remainingSources = sources
        
        while !remainingSources.isEmpty {
            let selected = selectByWeight(remainingSources)
            remainingSources.removeAll { $0.id == selected.id }
            
            let result = await executeWithTimeout(selected, loader: loader)
            recordResult(selected.id, result: result)
            
            if result.isSuccess {
                return result
            }
        }
        
        return .noFill(sourceId: "priority_group", reason: "No fill from priority group")
    }
    
    private func selectByWeight(_ sources: [WeightedAdSource]) -> WeightedAdSource {
        guard sources.count > 1 else {
            return sources[0]
        }
        
        let effectiveWeights = sources.map { source -> (WeightedAdSource, Double) in
            (source, calculateEffectiveWeight(source))
        }
        
        let totalWeight = effectiveWeights.reduce(0) { $0 + $1.1 }
        guard totalWeight > 0 else {
            return sources.randomElement()!
        }
        
        var random = Double.random(in: 0..<totalWeight)
        for (source, weight) in effectiveWeights {
            random -= weight
            if random <= 0 {
                return source
            }
        }
        
        return sources.last!
    }
    
    private func calculateEffectiveWeight(_ source: WeightedAdSource) -> Double {
        let baseWeight = source.weight
        
        guard config.usePerformanceWeighting else {
            return baseWeight
        }
        
        lock.lock()
        let stats = sourceStats[source.id]
        lock.unlock()
        
        guard let stats = stats, stats.attempts >= config.minSampleSize else {
            return baseWeight
        }
        
        let fillRate = Double(stats.successes) / max(1.0, Double(stats.attempts))
        let performanceMultiplier = 0.5 + fillRate
        
        return max(0.1, baseWeight * performanceMultiplier)
    }
    
    private func executeWithTimeout(
        _ source: WeightedAdSource,
        loader: @escaping AdLoader
    ) async -> AdLoadResult {
        let timeout = config.adaptiveTimeoutsEnabled 
            ? calculateAdaptiveTimeout(source) 
            : source.timeout
        
        return await withTaskGroup(of: AdLoadResult?.self) { group in
            group.addTask {
                do {
                    return try await loader(source)
                } catch {
                    return .error(sourceId: source.id, error: error)
                }
            }
            
            group.addTask {
                try? await Task.sleep(nanoseconds: UInt64(timeout * 1_000_000_000))
                return nil
            }
            
            if let firstResult = await group.next() {
                group.cancelAll()
                return firstResult ?? .timeout(sourceId: source.id)
            }
            
            return .timeout(sourceId: source.id)
        }
    }
    
    private func calculateAdaptiveTimeout(_ source: WeightedAdSource) -> TimeInterval {
        lock.lock()
        let stats = sourceStats[source.id]
        lock.unlock()
        
        guard let stats = stats, stats.successes >= config.minSampleSize else {
            return source.timeout
        }
        
        let avgLatencyMs = Double(stats.totalLatencyMs) / Double(stats.successes)
        let adaptiveTimeoutMs = avgLatencyMs * 2
        
        return min(source.timeout, max(1.0, adaptiveTimeoutMs / 1000))
    }
    
    private func recordResult(_ sourceId: String, result: AdLoadResult) {
        lock.lock()
        defer { lock.unlock() }
        
        let stats = sourceStats[sourceId] ?? SourceStats()
        stats.attempts += 1
        
        switch result {
        case .success(_, _, let latencyMs, let bid):
            stats.successes += 1
            stats.totalLatencyMs += latencyMs
            stats.totalBidMicros += Int64(bid * 1_000_000)
            stats.lastSuccessTime = Date()
        case .noFill:
            stats.noFills += 1
        case .error:
            stats.errors += 1
        case .timeout:
            stats.timeouts += 1
        }
        
        sourceStats[sourceId] = stats
    }
}

// MARK: - WeightedAdSourceBuilder

/// Builder for creating weighted ad source configurations
public final class WeightedAdSourceBuilder {
    private var id: String = ""
    private var priority: Int = 0
    private var weight: Double = 1.0
    private var timeout: TimeInterval = 5.0
    private var enabled: Bool = true
    private var minBid: Double = 0.0
    private var metadata: [String: String] = [:]
    
    public init() {}
    
    @discardableResult
    public func id(_ id: String) -> Self {
        self.id = id
        return self
    }
    
    @discardableResult
    public func priority(_ priority: Int) -> Self {
        self.priority = priority
        return self
    }
    
    @discardableResult
    public func weight(_ weight: Double) -> Self {
        self.weight = weight
        return self
    }
    
    @discardableResult
    public func timeout(_ timeout: TimeInterval) -> Self {
        self.timeout = timeout
        return self
    }
    
    @discardableResult
    public func enabled(_ enabled: Bool) -> Self {
        self.enabled = enabled
        return self
    }
    
    @discardableResult
    public func minBid(_ minBid: Double) -> Self {
        self.minBid = minBid
        return self
    }
    
    @discardableResult
    public func metadata(_ key: String, _ value: String) -> Self {
        self.metadata[key] = value
        return self
    }
    
    public func build() -> WeightedAdSource {
        precondition(!id.isEmpty, "Source ID is required")
        return WeightedAdSource(
            id: id,
            priority: priority,
            weight: weight,
            timeout: timeout,
            enabled: enabled,
            minBid: minBid,
            metadata: metadata
        )
    }
}
